pkg/module/oci/client.go

```diff
@@ -81,9 +92,117 @@ func (c *client) Cache() Cache {
+	}
+	ref = refspec.String()
+
+	resolver := c.resolver()
+
+	_, desc, err := resolver.Resolve(ctx, ref)
+	if err != nil {
+		return ocispecv1.Descriptor{}, nil, err
+	}
+
+	if desc.MediaType == MediaTypeDockerV2Schema1Manifest || desc.MediaType == MediaTypeDockerV2Schema1SignedManifest {
+		convertedManifestDesc, err := ConvertV1ManifestToV2(ctx, c, c.cache, ref, desc)
+		if err != nil {
+			return ocispecv1.Descriptor{}, nil, fmt.Errorf("unable to convert v1 manifest to v2: %w", err)
+		}
+		desc = convertedManifestDesc
+	}
+
+	if !isSingleArchImage(desc.MediaType) && !isMultiArchImage(desc.MediaType) {
+		return ocispecv1.Descriptor{}, nil, fmt.Errorf("media type is not an image manifest or image index: %s", desc.MediaType)
+	}
+
+	data := bytes.NewBuffer([]byte{})
+	if err := c.Fetch(ctx, ref, desc, data); err != nil {
+		return ocispecv1.Descriptor{}, nil, err
+	}
+	rawManifest := data.Bytes()
+
+	return desc, rawManifest, nil
+}
+
 func (c *client) GetManifest(ctx context.Context, ref string) (*ocispecv1.Manifest, error) {
-	// TODO
-	return nil, nil
+	desc, rawManifest, err := c.GetRawManifest(ctx, ref)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get manifest: %w", err)
+	}
+
+	if desc.MediaType != ocispecv1.MediaTypeImageManifest && desc.MediaType != images.MediaTypeDockerSchema2Manifest {
+		return nil, fmt.Errorf("media type is not an image manifest: %s", desc.MediaType)
+	}
+
+	var manifest ocispecv1.Manifest
+	if err := json.Unmarshal(rawManifest, &manifest); err != nil {
+		return nil, fmt.Errorf("unable to unmarshal manifest: %w", err)
+	}
+
+	return &manifest, nil
+}
+
+func (c *client) Fetch(ctx context.Context, ref string, desc ocispecv1.Descriptor, writer io.Writer) error {
+	refspec, err := ParseRef(ref)
+	if err != nil {
+		return fmt.Errorf("unable to parse ref: %w", err)
+	}
+	ref = refspec.String()
+
+	reader, err := c.getFetchReader(ctx, ref, desc)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := reader.Close(); err != nil {
+			c.logger.Error(err, "failed closing reader", "ref", ref)
+		}
+	}()
+
+	if _, err := io.Copy(writer, reader); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *client) getFetchReader(ctx context.Context, ref string, desc ocispecv1.Descriptor) (io.ReadCloser, error) {
+	if c.cache != nil {
+		reader, err := c.cache.Get(desc)
+		if err != nil && !errors.Is(ErrNotFound, err) {
+			return nil, err
+		}
+		if err == nil {
+			return reader, nil
+		}
+	}
+
+	resolver := c.resolver()
+
+	fetcher, err := resolver.Fetcher(ctx, ref)
+	if err != nil {
+		return nil, err
+	}
+	reader, err := fetcher.Fetch(ctx, desc)
+	if err != nil {
+		return nil, err
+	}
+	// try to cache
+	if c.cache != nil {
+		if err := c.cache.Add(desc, reader); err != nil {
+			// do not throw an error as cache is just an optimization
+			if err = reader.Close(); err != nil {
+				c.logger.Info("unable to close reader", "ref", ref, "error", err.Error())
```

Comment:
Should this be `c.logger.Error` like line 161?

Reply:
it should not treat as error, just tell the reader can't close, add to cache failed, but this function still can achieve by using `fetcher.Fetch(ctx, desc)`, so info level log is fine here.

Reply:
I see, thanks for clarifying.